File: 117-populating-next-right-pointers-in-each-node-ii/117-populating-next-right-pointers-in-each-node-ii.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* left;
    Node* right;
    Node* next;

    Node() : val(0), left(NULL), right(NULL), next(NULL) {}

    Node(int _val) : val(_val), left(NULL), right(NULL), next(NULL) {}

    Node(int _val, Node* _left, Node* _right, Node* _next)
        : val(_val), left(_left), right(_right), next(_next) {}
};
*/

class Solution {
public:
    Node* connect(Node* root) {
        
        if (!root) return root;
        
        Node *P, *parent;
        
        root->next = nullptr;
        parent = root;
        P = nullptr;
        Node* prev = nullptr;
        
        while (parent) {
            
            if (parent->left && parent->right) {
                if (prev) prev->next = parent->left;
                else P = parent->left;
                
                parent->left->next = parent->right;
                prev = parent->right;
            }
            else if (parent->left) {
                if (prev) prev->next = parent->left;
                else P = parent->left;
                prev = parent->left;
            }
            else if (parent->right) {
                if (prev) prev->next = parent->right;
                else P = parent->right;
                prev = parent->right;
            }
            parent = parent->next;

            if (!parent) {
                if (!P) return root;
                
                parent = P;
                prev = P = nullptr;
            }
        }
        
        return root;
    }
};

File: 117-populating-next-right-pointers-in-each-node-ii/117-populating-next-right-pointers-in-each-node-ii.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""
class Solution:
    def connect(self, root: 'Node') -> 'Node':
        
        if not root:
            return root
        
        root.next = None
        nodes = [root]
        cnt = 1
        
        while cnt:
            
            nextCnt = 0
            
            for idx in range(cnt):
                
                node = nodes[0]
                nodes.pop(0)
                
                if node.left:
                    if nextCnt:
                        nodes[-1].next = node.left
                    nodes += [node.left]
                    nextCnt += 1
                    
                if node.right:
                    if nextCnt:
                        nodes[-1].next = node.right
                    nodes += [node.right]
                    nextCnt += 1
            
            cnt = nextCnt
        
        return root
                
            
            
            
            
            
            
        
        
            
            
        
        
        
        